### DISCUSSION TITLE
Variable names starting with _underscores

### DISCUSSION THREAD

**Discussion by EduardoVB** [twinbasic/twinbasic#1051] [Polls]
VBx does not allow variable names starting with underscores, other languages do.
tB currently allow that.

In my opinion that's not BASIC-ist. The code readability is greatly diminished.
Why to allow it, is there any reason? (other than copying what other languages do)

**Comment by wqweto** [comment]
Technically it's already allowed in VB6 with enum entries.
```
Option Explicit

Private Enum MyEnum
    [_What]
End Enum


Private Sub Form_Load()
    Debug.Print [_What]
End Sub
```
As far as I know TB allows quoted identifiers (the `[` and `]`) in a lot of more places e.g. this
```
    Private Sub [My Sub]()
            	
    End Sub
```

**Comment by bclothier** [comment]
I find it useful for denoting private variables and not have to invent yet another name or use prefixes:

```
Private _Foo As String
Public Property Get Foo() As String
  Return _Foo
End Property
Public Property Let Foo(NewValue As String)
  _Foo = NewValue
End Property
```

It's less typing than `m_Foo`, eliminates the need to remember if it's `g_Foo` or `m_Foo`. I've had enough of codebases where we have `Public m_intFoo As String` or `Private g_intFoo As String`. Those prefixes do not make refactoring easy and they just encourage more cruft.

So this feature would help us move away from the [backward System Hungarian notations](https://www.joelonsoftware.com/2005/05/11/making-wrong-code-look-wrong/) which is deeply harmful to maintainability.

**Comment by EduardoVB** [comment]
I now that you guys love these things, but it is because you are already contaminated with .Net stuff.

It is ugly, not Basic-ish.
I don't know how to emphasize more.
It ruins the legibility.

You could use the same hackneyed argument: if you don't like then don't use it.
That argument could be used to introduce any atrocity to the language, then you'll find lot of tB code on internet and that will be a pain to read. That's not the spirit of Basic.

Please disallow that. VB never had that, why to introduce something (BTW unneeded) like this?
That a couple of persons with .NET background think it is a good idea does not seem enough reason IMO.

**Reply by mansellan** [reply]
It's not really about .Net though, plenty of people choose not to use underscores there, and just differentiate properties from class-level variables using `camelCasing` for variables. That's fine because C# is case sensitive, but VB is not.

So the alternative is to use a prefix like `m_`, which just adds noise (to my eye at least).

That said, I'm really not massively bothered either way - if people want them outlawed then that's cool too.

**Comment by FullValueRider** [comment]
True.  I think the definitive solution to this issue was described in a Ruber duck blog.

https://rubberduckvba.wordpress.com/2018/04/

I didn't like the use of 'this' as I felt it was  a transplant for C#, but I do like using two UDT, which I call State and Properties used with variables s and p, which encapsulate module level variables used internally (State) and module level variable which are exposed as  Properties.

I also got fed up with name clashes so adopted the following 4 prefixes
my - for variables declared within a Method
ip - for a parameter that supplies a value
op - For a parameter that exists solely to pass a result back to the caller
iop - for a parameters that will be mutated by the method.

It all works rather splendidly well.

**Reply by EduardoVB** [reply]
@ FullValueRider then you are against the prefix?
But I don't see your vote.

**Comment by EduardoVB** [comment]
Once more: VBx is a language not intended (mainly) for geeks/nerds but for normal people.
No one in their right mind would name something starting with _ because that cannot be pronounced, and cannot be read normally.

Different is the case of the underscores at the middle, they act as separator, as a replacement of the space that you are not allowed to use there. That's a total different reason.

If you want to use a character to mean something, some hungarian notation, use another character.
C people are not normal people, do not copy their behaviors.

Do you think that the VB6 original developers disallowed it for not reason, just because they were stupid (and you are much smarter)?

About the += you had some somewhat reasonable argument (well, really the only one was the terseness), about this one I don't see any reason to ruin the code like this. It becomes unreadable.
Why do you think why many of us program in VB and not in C?
Do you really think that most people will come from C??????

**Reply by mburns08109** [reply]
> C people are not normal people, do not copy their behaviors.

ROFL!

**Comment by EduardoVB** [comment]
We are sometimes influenced by our world we are used, but for other people, I'm mean for people still not influenced by our little wold, some things are weird.
Anyone that comes from the outside world will see as weird words starting with _ because they cannot be pronounced.

We all suffer from this kind of things that appear normal to us but look weird to others.
For example I remember in a program I wrote an option "Copy to clipboard".
Then, some users that were testing (normal people) asked "what is clipboard?" and they asked to change that weirdness to just "Copy". Of course they were right! 
We need to be open, not to think that the way we see things is the only one or the best.
I believe the original VB developers were not fouls.

**Comment by wqweto** [comment]
Btw, potential converts writing in C are 100x more that ones using VB6/VBA. Even more people are using C# or TypeScript daily in the wide open world of non-VB programming languages.

**Comment by mansellan** [comment]
Can we have another voting option please?

- Whichever option doesn't lead to another 100-comment thread :-)

**Reply by mburns08109** [reply]
(calls out) MAN your Chariots! We have a new religious war to fight! (/calls out)

Uh... uhm... What's our rally cry to be? are we the "Save the Underscore" champions, or are we the "Supress the evil, demonic underscores!" side? I forget.

**Comment by FullValueRider** [comment]
> @ FullValueRider then you are against the prefix?
> But I don't see your vote.

I don't see this as an issue worthy of needing a vote.  The use of the underscore is not mandatory so it doesn't affect my writing of code.

As for reading of code, legacy VBA code is a sufficiently bad actor that having to deal with a preceding \_ now and again is only a small increment in the overall terribleness.

I'd only be concerned if the preceding \_ started causing issues with the line continuation facility of VBA, or if it compromised the introduction of a new and significantly better way of doing something else in VBA.

What we must be cognisant of is that many things in relation to how code is written represent pragmatic decisions on how best to achieve a goal at that point in time.  These things are not necessarily what we would consider best practise.  However, as time has moved on, compilers have become cleverer, memory has become less constraining, the importance a clean code/transparency over clever code/ overly concise code has become increasingly obvious.

If we want VBA to run legacy code we have to accept stuff that we wouldn't now implement.  If twinBasic wants to win over coders from other paradigms it shouldn't shut out established practise that can be transferred from those paradigms.  Instead, there should be a set of community agreed guidelines that point out how to write good code in VBA/twinbasic.

That's why I made the reference to the Rubberduck blog article and explained why I do things differently (but following the lead of the Rubberduck article).

Sorry if this disappoints you.

**Reply by FullValueRider** [reply]
You can see that I'm not that tech savvy as I didn't realise I had to escape the  underscore.

**Comment by fafalone** [comment]
You don't have to use it, if you don't want to, don't. More flexibility is almost always better.

I don't personally like them, because why require an arbitrary character you need shift to put in? But then, I bring in a lot of C/C++ code to VB, and I also hate having to remove them from everything. As long as it's not breaking line continuations, I say keep it. I'm not naming my own variables that way, I'd imagine most people who aren't geeks/nerds won't be either. But nothing is forcing them to. I guess I just don't see as big a difference between the fact we already make extensive use of them as space replacements and sticking one at the start if you really want to.

**Comment by mansellan** [comment]
I'd suggest this belongs in lang-design, because it's not a back-compat issue, just the removal of an artificial restriction in VBx.

**Comment by WaynePhillipsEA** [comment]
I'm generally on the side of not adding artificial limitations unless necessary.  However, with not being a normal person, I will listen to the community on this 😉

**Comment by VBFNiya** [comment]
I voted to allow. I prefix class level variables with underscores in my own VB.Net code. However, this is not a hill I'd die on. In VB6, I use the prefix "g_" or "m_" for class/module level variables. It's only one extra character. I prefer using just the underscore but I don't really mind having to type one extra character if I absolutely have to. In fact, I sometimes out of sheer habit from years spent in VB6 still us the "g_" prefix in .Net code.

**Comment by mansellan** [comment]
It seems that there isn't universal hatred towards leading underscores, so I'd suggest we don't outlaw them.

@WaynePhillipsEA does that seem reasonable?

**Reply by WaynePhillipsEA** [reply]
Yeah, I'd agree with that.

**Comment by rexxitall** [comment]
I hate them cause they are hard to read and will destroy backwart compatibility 
![image](https://user-images.githubusercontent.com/22050310/185741755-173357e8-5e94-479f-9eee-af99fd03bb40.png)

**Reply by rexxitall** [reply]
If allowed they will be used for sure. If you want to have a great eco system then you should avoid such things. The code should be run without changes in TB, VBA, VB6. It will be frustratin if some copy twinbasic code in VBA just to see how compatible it is and get hundreds of error messages. Maybe he will not buy it. Same reason why i suggest to start compiler statements like [dllexport] with this >'< char to mark them as comment or at least allow it to avoid #if #endif orgys. :)

**Reply by WaynePhillipsEA** [reply]
It would be way too restrictive to not be allowed to include any new syntax to support new features in the language.  The decision was made long ago that we need **_not_** be concerned with forward-compatibility (i.e. ensuring all tB code can be copied into VBx).  There is no movement on that issue, as we are way too far along for that.

**_Backwards-compatibility_** with VBx is the primary goal, not the other way around.

**Reply by mwolfe02** [reply]
@rexxitall For the use case you are describing, there was talk about adding an `Option Legacy` (or similar syntax) that would guarantee forward-compatibility at the module and/or project level.  I'm not sure whether any decision was made to support that or not (or where it would appear on the tB Roadmap).

**Comment by fafalone** [comment]
VB/VBA versions have never had forward compatibility with eachother. Nor has Windows. 

You don't get new versions of things and successors because you want to be limited to the exact feature set of the old thing. What even would be the point? 

And what would you want to be compatible with? Lots of VBA stuff will throw hundreds of errors if copied into VB6.

Lots of VBA7x64 stuff will throw hundreds of errors if copied into VBA6x86.

If somebody wants to be compatible with the older languages, they can do what VBA programmers do. `#If VBA7 Then` `Else`. 

What I would recommend is a built in compiler flag to indicate twinBASIC is present; that way it's easy for people who do want code compatible in all products to create separate versions just like for people who want VBA7/VBA6/VB6 compatibility all in the same project.

```
#If TWINBASIC Then
Public DeclareWide PtrSafe Function ...
#ElseIf VBA7 Then
Public Declare PtrSafe Function
#Else
Public Declare Function
#End If
```

**Reply by WaynePhillipsEA** [reply]
Yup, that compiler constant already exists 😉
